app/src/behaviors/behavior_hold_tap.c

```diff
@@ -370,19 +374,40 @@ static void decide_hold_tap(struct active_hold_tap *hold_tap,
         return;
     }
 
-    switch (hold_tap->config->flavor) {
-    case FLAVOR_HOLD_PREFERRED:
-        decide_hold_preferred(hold_tap, decision_moment);
-    case FLAVOR_BALANCED:
-        decide_balanced(hold_tap, decision_moment);
-    case FLAVOR_TAP_PREFERRED:
-        decide_tap_preferred(hold_tap, decision_moment);
+    // For conditional hold-tap behaviors that have failed to meet the required prerequisites
+    // for a hold decision, force a tap decision.
+    if (decision_moment == HT_OTHER_KEY_DOWN && hold_tap->config->hold_enabler_keys_len > 0) {
```

Comment:
I don't think this if statement is correct for all 'flavors'. For example, the `balanced` never decides on HT_OTHER_KEY_DOWN; it decides on HT_OTHER_KEY_UP. I think we should put this code into a function `is_hold_enabled_key(int32 other_key_decision, struct active_hold_tap *hold_tap)`, and call it from each of the `decide_*` methods.

Somewhat tricky would be the behavior of `tap-preferred`. I think it should act like this:

* if the timer has not run out (e.g., we have not seen `HT_TIMER_EVENT`, always trigger tap. (doesn't matter which key is pressed)
* if the timer did run out (the HT_TIMER_EVENT was seen, maybe setting STATUS_UNDECIDED_TIMER_SEEN), trigger HOLD if the key is a hold_enabler_key, otherwise trigger TAP

For balanced, the decision_moment to check for hold_enabler_keys would be on HT_OTHER_KEY_UP.

Reply:
you're right, I'll fix. I'll also make the testing more robust to catch this.

Reply:
I think we'll still need to store position_of_first_other_key_pressed in the active_hold_tap struct, because in tap-preferred situation 4a, I don't see any other way of verifying whether the first other key that was pressed was an enabler key, especially if a second key is pressed before tapping term expires

![image](https://user-images.githubusercontent.com/44815547/133153781-0b749af7-80b7-42d2-9d43-30e2b43f55d3.png)